---
title: "parrot.pca"
author: "Devon DeRaad"
date: "12/14/2020"
output: html_document
---

```{r, message=F, warning=F}
#load libraries
library(vcfR)
library(adegenet)
library(ggplot2)
library(introgress)
library(StAMPP)
```

```{r}
#load data
vcf<-read.vcfR("/Users/devder/Desktop/parrots/la.parrots/parrotsnpsunfiltered.vcf")
#convert vcfr to genlight
gen<- vcfR2genlight(vcf)
#read in locality info for samples
locs<-read.table("~/Desktop/parrots/la.parrots/parrotsamples.txt", header=T, sep="\t")
colnames(locs)<-c("id","location","tiss")
locs$group<-as.character(locs$location)
locs$group[3:6]<-"Lilac Mexico"
locs$group[27:30]<-"Red Crown Mexico"
gen@ind.names

#check that sampling IDs in text file match sample IDs in the vcf
gen@ind.names == locs$id
```


```{r}
#calculate missingness per sample
mat<-extract.gt(vcf)
miss<-c()
for (i in 1:ncol(mat)){
  miss[i]<-sum(is.na(mat[,i]))
}
locs$miss<-miss
locs
```
elevated number of missing genotypes in toepad samples, as expected

filter genlight to 60, 80, 100 % complete matrices
```{r}
#make vector to store missing proportion of each SNP
miss<-c()
gen.mat<-t(as.matrix(gen))
for (i in 1:nrow(gen.mat)){
  miss[i]<-sum(is.na(gen.mat[i,]))/ncol(gen.mat)
}

gen.60<-gen[,miss < .4]
gen.80<-gen[,miss < .2]
gen.100<-gen[,miss == 0]

```

#PCA with 8508 SNPs passing 60% completeness cutoff
```{r}
pca<-glPca(gen.60, nf=6)

#pull pca scores out of df
pca.scores<-as.data.frame(pca$scores)
pca.scores$group<-locs$group

#ggplot color by species
ggplot(pca.scores, aes(x=PC1, y=PC2, color=group)) +
  geom_point(cex = 5, alpha=.5)+
  scale_color_manual(values=c("green","blue","red"))+
  theme_classic()

```

#PCA with 3413 SNPs passing 80% completeness cutoff
```{r}
pca<-glPca(gen.80, nf=6)

#pull pca scores out of df
pca.scores<-as.data.frame(pca$scores)
pca.scores$group<-locs$group

#ggplot color by species
ggplot(pca.scores, aes(x=PC1, y=PC2, color=group)) +
  geom_point(cex = 5, alpha=.5)+
  scale_color_manual(values=c("green","blue","red"))+
  theme_classic()
```

#PCA with 1070 SNPs passing 100% completeness cutoff
```{r}
pca<-glPca(gen.100, nf=6)

#pull pca scores out of df
pca.scores<-as.data.frame(pca$scores)
pca.scores$group<-locs$group

#ggplot color by species
ggplot(pca.scores, aes(x=PC1, y=PC2, color=group)) +
  geom_point(cex = 5, alpha=.5)+
  scale_color_manual(values=c("green","blue","red"))+
  theme_classic()
#best to stick with no missing data in order to avoid issues of missing data influencing group assignment

#ggplot with text labels
ggplot(pca.scores, aes(x=PC1, y=PC2))+
  ggrepel::geom_text_repel(label=locs$id, cex=3)+
  xlim(-10, 5)+
  theme_classic()
```

```{r}
### Calculate Nei's distances between individuals/pops
pop(gen.100)<-locs$group
gen.100@ind.names<-substr(locs$id, 7,14)
sample.div <- stamppNeisD(gen.100[c(1:6),], pop = FALSE)
#lilac NJ tree
plot(nj(sample.div), type = "unrooted", cex = .65)
sample.div <- stamppNeisD(gen.100, pop = FALSE)
#export for splitstree
stamppPhylip(distance.mat=sample.div, file="~/Downloads/parrot.splits.txt") 
#all samples splitstree
knitr::include_graphics(c("/Users/devder/Desktop/parrot.splitstree.png"))
#seems like kind of equivocal placement of the LA lilac crown samples. The divergence within lilac crowned is quite shallow.
gen.lilac<-gen.100[c(3:6),]
gen.lilac@pop<-as.factor(c("n.mex","s.mex","n.mex","s.mex"))
stamppFst(gen.lilac, nboots = 1, percent = 95, nclusters = 1)
#Fst .05 between the two ancestral lilac crowned subspecies

gen.rc<-gen.100[c(7:23,25,27:30),]
gen.rc@pop<-as.factor(c("lower","upper","lower","lower","upper","lower","lower","upper","lower","lower",
                           "lower","upper","lower","upper","upper","upper","upper","lower","lower","lower","lower","lower"))
stamppFst(gen.rc, nboots = 1, percent = 95, nclusters = 1)
#Fst .13 between red crowned clusters

#actually looks like a deeper split within red-crowned than within lilac crowned
#all of our historical red crowned samples are in the bottom cluster
#top cluster could be from a different part of the historical range, or from a different contemporary source (Texas or Florida established pops?)
#Do we see two mitochondrial clades within our red-crowned sample?
#Could also be explained by ghost introgression, alleles introgressed from some species we didn't sample here generating a clade within the LA red crowned pop.

#Also this splitstree is a nice visualization supporting that those three red crowned samples in particular (B728, B726, and B564) with the most intermediate PC1 scores, being advanced backcrosses.
```

```{r}
#read in csv with estimated hybrid index
hi<-read.csv("~/Downloads/parrot.hybrid.index.csv")

pca.scores$hi<-hi$h
plot(x=pca.scores$PC1, y=pca.scores$hi,
     xlab="PC1 with no prior assignment", ylab="hybrid index based on fixed SNPs in parental pops",)
#PC1 for all samples is highly correlated with hybrid index
#amongst the LA red-crowned samples, we see a correlation btwn hybrid index score and PC1 assignment,  
#indicating that these these samples with the largest hybrid index score are also more like 
#lilac crowned according to the PCA which is non-parametric and requires no a priori assignment. 

#The LA lilac crowned samples are not any more red-crowned than the Mexico lilac crowned samples according to PCA

#It seems like there are clearly some real advanced backcrosses (I see three in particular that have higher hybrid indexes from introgress, and are slightly shifted toward lilac crown in PC1) in the LA red crowned population, but for all intents and purposes the LA pop is all parental individuals with 2 hybrids. There are red crowned alleles that are being lost via introgressive hybridization, but there are still a good amount of LA samples that could be considered pure parental stock for conservation purposes.

pca<-glPca(gen.100[c(1,2,7:26),], nf=6)
hist(pca$scores[,1])
plot(density(pca$scores[,1]), xlab="PC1")
#cool to see that this conforms almost exactly to the expectations of a bimodal hybrid population

#calculate Fst between the two parental populations
gen.ancestral<-gen.100[c(3:6,27:30),]
gen.ancestral@pop<-as.factor(c(rep("lc", times=4),rep("rc", times=4)))
stamppFst(gen.ancestral, nboots = 1, percent = 95, nclusters = 1)

#calculate Fst between the two phenotypic classes in LA
gen.la<-gen.100[c(1,2,7:23,25),]
gen.la@pop<-as.factor(c(rep("lc", times=2),rep("rc", times=18)))
stamppFst(gen.la, nboots = 1, percent = 95, nclusters = 1)
#if you remove the two hybrid individuals, Fst between the two phenotypic classes in LA is actually higher than between the ancestral stock from Mexico, but only two LC samples from LA means that this is probably not significantly greater, probably well within confidence intervals with such small sample sizes. Still, evidence that overall divergence between the lineages is being strongly maintained despite hybridization.
```


